### DISCUSSION TITLE
The largest _w90_tolerance_

### DISCUSSION THREAD

**Discussion by Lanlin123** [TRIQS/solid_dmft#75] [Q&A]
Hi everyone,
    When I did a CSC caculation, I met the error shown as the following figure. Here I have a question that how I choose the largest _w90_tolerance_ to get reasonable result.
![image](https://github.com/TRIQS/solid_dmft/assets/51743630/d9c3551b-fd29-4f7f-9a46-70cdd836d177)

**Comment by Lanlin123** [comment]
Besides, in my calculation, _band_energy_modscf_ defined in code file _qe_manager.py_  can't be found in _mod_sef.out_ file as there is not a line including "(sum(wg*et))". Could you give me some advice? Thanks!

**Comment by phibeck** [comment]
Hi,
the error tells you that two shells that are equivalent at the ground state level (i.e. at the beginning of your calculation) start to become numerically inequivalent during the CSC workflow.

To avoid the error, you can in principle lower the tolerance in the input file in the `[advanced]` parameters section, in your case you could set `w90_tolerance = 6e-3`. In practice, what you should do is increase the accuracy of your calculation to make sure that the errors introduced by the mapping of sites don't grow too large. For this I would recommend increasing the accuracy of your impurity solver (in case it is CT-HYB, here are some guidelines https://triqs.github.io/cthyb/latest/guide/cthyb_convergence_tests.html) and the number of k-points.

Regarding your second question, I believe this will be fixed if you reinstall the latest version of Quantum Espresso. @alberto-carta could you comment on this?

**Reply by alberto-carta** [reply]
Dear LanLin, could you please attach to the issue the input files for the mod_scf calculation, the standard output and error of solid_dmft? What comes to mind regarding your problem could be:

* You are not using the newer version of quantum espresso, as mentioned by Sophie (>7.1 for the new fixes to the DMFT interface, see this commit on the QE gitlab https://gitlab.com/QEF/q-e/-/commit/7f61b53e51dd85f7097cd5ecf04e2ddd953bd178)
* You may have forgot to include dmft =.true. in the &system card in the mod_scf input file
* the calculation failed for another reason, but solid_dmft still finds the mod_scf.out file in the folder and tries to read from it, I recommend to delete all the *out, temporary folder and h5 to give everything a fresh start.

**Reply by Lanlin123** [reply]
Hi, Phibeck and Alberto-carta:
    Thank you so much for your reply. I have taken your advices and they indeed work. Though I met another error after the first DMFT iterations in my CSC calculation as shown below. Could you give me some advice? Thanks!
![image](https://github.com/TRIQS/solid_dmft/assets/51743630/2f0add8c-7cfa-4916-8497-dee6e25b305e)

**Reply by alberto-carta** [reply]
Dear LanLin, could you please provide the input files?

**Reply by Lanlin123** [reply]
Dear Alberto, please see [https://github.com/Lanlin123/FeSe](url) . The error occured between two DMFT cycles in CSC calculation.

**Reply by Lanlin123** [reply]
Please copy the link and access to it in a new window.

**Reply by phibeck** [reply]
Hi @Lanlin123, this seems to be a bug in the Wannier90Converter. I've opened an issue in https://github.com/TRIQS/solid_dmft/issues/76. We'll try to address this as soon as possible, thanks for providing input files.

**Reply by Lanlin123** [reply]
Thanks for your works! Could you please give me a message after you address it. Thank you again.

**Reply by Lanlin123** [reply]
Hi Phibeck and Alberto,
    The mismatch between downfolded Hamiltonian and Fourier transformed H(R) is very large as shown below. Could you please give some comments on it?
![image](https://github.com/TRIQS/solid_dmft/assets/51743630/9a279df7-a86c-4e1f-b9d1-c23c6ba3e16e)

**Reply by phibeck** [reply]
Hi @Lanlin123, this looks like an issue with the Wannierization. It's possible your Wannierization parameters need to be adjusted in the charge self-consistent cycle, as bands move around. It is important to monitor the Wannier functions spread and inspect the band structure of the Wannier Hamiltonian to see whether it still reflects the bands you want to treat with DMFT. Can you have a look at the `seedname_hr.dat_itX` and compare them with the previous iterations?

**Reply by Lanlin123** [reply]
Hi Phibeck, Thank you so much. I have adjusted the Wannierization parameters and it works. By the way, I wonder whether there is a Multi-Orbital Iterated-Perturbation Theory sovler for solid_dmft.

**Reply by the-hampel** [reply]
Hi @Lanlin123 , 
Let me quickly jump in here an answer your question regarding solver: As of now we do not have a general IPT solver within TRIQS. The only cheap solvers that are currently available are https://triqs.github.io/hubbardI and https://triqs.github.io/hartree_fock/ . Those are also available in solid_dmft. Let me know if you have questions regarding those.

**Reply by Lanlin123** [reply]
Thank you for your reply! I have calculated FeSe and get a weird self energy as shown below. The block 0 component seems reasonable comparing to the previous work corresponding to dxy orbital. While the others deviate a lot. The files can be accessed from https://github.com/Lanlin123/FeSe. Can you give me some advice? 
![image](https://github.com/TRIQS/solid_dmft/assets/51743630/bd3bd508-8646-4fc1-8f98-d8cf7e6f20c9)
The following picture shows results from the previous work.
![image](https://github.com/TRIQS/solid_dmft/assets/51743630/0cee2f99-573f-4ac5-a0f2-458419a9d85b)

**Reply by the-hampel** [reply]
I had a quick look at your analytic continuation (AC) and plotting notebook. I think there is a small mistake in the plotting script. I think you wanted to subtract the real part of Sigma(w=0.0) to make curves better comparable. But you also substracted the imag part (both), which is a bit confusing I think for looking at the self-energy qualtiy. Can you replot only substracting the real part: 
```
for block, Sw in Sigma_real_freq:
    ax.oplot((Sw - Sw(0.0).real)[0,0].imag ,'-', c='C1', label=block)
```
However, I also see that you observe a pole in S(w) just below 0, which seems to be absent in the reference data. May I ask you to plot or provide the Matsubara self-energy as well that you continued (Im Sigma (iwn)) is fine. They are also stored in the h5 archive of solid_dmft under `DMFT_results/last_iter/Sigma_freq_0` . 

Best,
Alex

**Reply by Lanlin123** [reply]
Hi Alex,
Thanks for your reply. I plot the Matsubara self-energy shown below. The plotting notebook is upgraded in https://github.com/Lanlin123/FeSe.
![image](https://github.com/TRIQS/solid_dmft/assets/51743630/37ad1334-1e9e-4a0f-bcaa-604cd4830504)

**Reply by the-hampel** [reply]
I think now the plot of the AC self-energies looks okay. The Matsubara data looks also reasonable I would say. Since, this is paramagnetic calculation you could maybe only plot the up channel and skip plotting all down blocks. However, I would say the result looks at least physical to me. There is no obvious problem I see. I am not an expert on FeSe, but now we have to evaluate how similar the two setups between your reference plot you showed and this calculation, e.g. how similar are the local orbitals used, U / J values, structure. This I cannot judge. 

Are you using triqs/3.2.x? You could maybe improve accuracy a bit more by `measure_g_l = False` and instead do tail fitting by adding instead in the solver section of solid_dmft: 
```
perform_tail_fit = true
fit_max_moment = 4
fit_min_w = 3.0
fit_max_w = 8.0
```
Since version 3.2.x this might have an edge compared to legendre sampling since we use the high frequency moments of the self-energy to improve the tail fit. The window iwn=3 to iwn=8 should be good for this parameter setting, but could change when changing U or J. Also keep an eye on the reported `autocorrelation time` in the solver output and adjust the `length_cycle` accordingly (https://triqs.github.io/cthyb/latest/guide/cthyb_convergence_tests.html#1.-length_cycle) . I did not find the output from the solid_dmft run so I cannot judge how well these parameters are chosen here.

**Reply by Lanlin123** [reply]
Thank you, Alex. Yeah, I am using triqs/3.2.x. I will try again as you suggest. As there is a pole in S(w) just below 0, but the Matsubara data looks physically reasonable, I wonder whether the problem do come from analytic continuation process.

**Reply by Lanlin123** [reply]
Hi, I have check carefully and conclude that the problem should come from analytic continuation. The Matsubara self energy from previous work is shown below, which seems to be consistent with my result shown above. The only problem is the order of the five d orbitals. In my wannier90 calculation, I use [dxy,dyz,dz^2,dxz,dx^2-y^2]. The _block structure summary_, _Rotation matrices_, _Effective atomic levels_ and _DFT density matrix_ are shown below. As you can see, the _Rotation matrices_ of the two eq. correlated shells are not the same which is weird and they give [dxy,dx^2-y^2,dyz,dxz,dz^2], [dxy,dxz,dyz,dx^2-y^2,dz^2] of d orbital orders after the global-to-local transformation respectively. However, as bolck up_3 and up_4 are dengerate, they should correspond to dxz/dyz orbitals, which deviate from the rotation result. Could you please give me some advice on how to determine the orbital order of different blocks in SumK? I think I don't understand the global-to-local transformation well enough, but I don't find more details on it. Could you please give some notes on it? Thanks.
![image](https://github.com/TRIQS/solid_dmft/assets/51743630/3f5b83f8-6b64-4329-9c9e-e69d770571fe)
![image](https://github.com/TRIQS/solid_dmft/assets/51743630/f82251fc-499c-40e2-9317-aca2ea0dd843)

**Reply by the-hampel** [reply]
Hi @Lanlin123, 
I think this looks all quite reasonable to me so far and I guess it is a good sign that the Matsubara self-energies agree reasonably well. I am not sure I understand your question regarding the block structure entirely. Let me explain what is happening in the DFTTools wannier90 converter when you run it through solid_dmft. 

In principle you have various options on how the wannier90 converter handles multiple impurity sites and local rotations when creating the h5 archive readable by triqs/solid_dmft (https://triqs.github.io/dft_tools/unstable/_modules/triqs_dft_tools/converters/wannier90.html#Wannier90Converter) . When running solid_dmft + QE in CSC some choices have been made, namely `bloch_basis=True` and `rot_mat_type='hloc_diag'`. For multiple as equivalent flagged impurities this means: The DFT eigenvalues are stored as input for DMFT and the lattice Green function is constructed in this Kohn-Sham (KS) basis. The projectors are not trivial and will produce the projected low-energy Wannier Hamiltonian at each k-point. Now, the wannier90 converter checks by comparing the eigenvalues on each flagged equivalent impurity site if they are indeed equivalent (otherwise an error is thrown). Then it will find a basis transformation (norb x norb unitary matrix) for each site that will give you the same diagonal structure (sorted eigenvalues of Hloc). So the rotation matrices shown here (global to local) make sure that each site has the same local Hamiltonian (so that rho[0] and rho[1] are identical above). In this basis in which the non-interacting local Hamiltonian is diagonal the impurity problem is solved (Hint is rotated accordingly) for the first impurity. Then the rotation is applied backwards to create the self-energy for each site in their original by w90 given basis. The projectors are applied then backwards to give back the lattice Gf. 

Now, why the orbitals are flipped between site 0 and site 1 is probably a result of rotations in your unit cell that will related the two Fe sites by a symmetry operation that is not identity. You can check that by looking at the `seed_hr.dat` file at R=(0,0,0) comparing the entries between the first and second Fe site. And indeed the degenerate shells are detected indicate that the last two orbitals are identical, which looks reasonable from a numerical point of view (same eigenvalues and occupation). 

I am not sure I understand why you want to change the final order for the impurity? In principle this is irrelevant for you. Important is the if the order before the global-to-local trafo is applied is correct. The rest should be just numerical recasting without information loss. When you use the AC functionality from solid_dmft it is done in the impurity basis. No transformations are applied yet. 

Maybe you can let me know to which part you have explicit questions. Also you can maybe play around a little bit with the AC parameters to check how consistent the AC self-energies are for parameter changes. Try to reduce `n_points_maxent` to something like 100 and see if you can get similar results. Also check other DMFT iterations in the AC in case you have doubt that everything is okay there.

**Reply by Lanlin123** [reply]
Hi @the-hampel,
    Thanks for your reply. Now I have no question about the d orbital order of impurity. I will play around a little bit with the AC parameters as you suggest. In the last two days, I am playing with the Ce2O3 tutorial [https://triqs.github.io/solid_dmft/tutorials/Ce2O3_csc_w90/tutorial.html](url). I do one shot calculations and try to dope holes by reducing the electron density in seedname.inp file. However, when the electron density is chosen as 1.8, 0.2 smaller than the original 2.0, the impurity occupancy becomes nearly zero and the final self energy is zero too. The same things happen when the electron density is smaller. I wonder whether solid_dmft can get non-zero solution in the dmft self consistent process in these hole doping cases. Thanks again.

**Reply by the-hampel** [reply]
Hi @Lanlin123 , 
In tutorial the HubbardI solver is used. The Hubbard-I solver neglects completely the coupling between impurity and bath. Thus it only solves the impurity problem and is limited to integer filling solutions + some temperature smearing of the filling. Hence, doping is very hard to do with the hubbardI solver. What is the reported electron count in the impurity for the default n=2.0 after the solver run?

**Reply by Lanlin123** [reply]
Hi @the-hampel,
The reported electron count in the impurity is 1.0 for  the default n=2.0 after the solver run,which is reasonable. Although when I try to dope less holes or electrons, I can get some non-zero solution as shown below. The final impurity occupation of these calculation is close to 1.0 as you say, but  the DOS of f electron changes a lot.I wonder whether they are physically reasonable. By the way, except CTHYB, we have no other cheaper solver to treat the doping case, right?
![image](https://github.com/TRIQS/solid_dmft/assets/51743630/a462a75f-d075-47fc-9243-a814e94ca775)

**Reply by the-hampel** [reply]
Okay yes then using HubbardI might be not ideal here for non-integer fillings. You can try to use the Hartree-Fock solver: https://triqs.github.io/hartree_fock/latest/ , which is also available in solid_dmft. Here is an example config file using this solver: https://github.com/TRIQS/solid_dmft/blob/3.2.x/test/python/svo_hartree/dmft_config.ini . But HF might be less suitable for atomic like problems like f-states. But it works of course also with fractional fillings. If you are only considering density-density (ising) like interaction Hamiltonians, the segmentation picture solver triqs/ctseg might be a good option though. It is much faster than traditional cthyb, but limited to density-density type interactions. The repository is private but if this would help you in your research we can invite you to the project. The project is private because documentation is very limited. However, it works with solid_dmft quite well.

**Reply by Lanlin123** [reply]
Hi @the-hampel ,
Thanks so much! It would be great if I could join the triqs/ctseg project as I mostly choose density-density like interaction Hamiltonians to treat full d/f orbital shells. I wonder what I need to do to join it.

**Reply by the-hampel** [reply]
I sent you an invite to the project. Please install the `unstable` branch for compatibility with triqs/3.2.x . In solid_dmft choose `solver_type=ctseg` and use `legendre_fit=True` as solver option with an appropriate choice of `n_l` . There is no other tail-treatment available in this solver. If needed I can also put the tail-fit in place there.

**Reply by Lanlin123** [reply]
Hi @the-hampel,
Thank you so much for inviting me to ctseg project. I have install the unstable branch and test it using FeSe,Ce2O3 and SrVO3. I think there exsist some imperfections in ctseg unstable branch at the moment. Firstly, I have calculate more than 70 iterations and the chemical potential becomes 1342 now which is weird. The files for this can be accessed from https://github.com/Lanlin123/test_for_ctseg.
Secondly, when I use it to Ce2O3 and SrVO3, it crashed due to insuffcient memory as shown below. As it only need to treat a five-orbital problem for SrVO3, which is simpler than FeSe's ten-orbital problem, I have no idea why it need more memory resources for SrVO3 than FeSe.
What's more? I have checked the FeSe results using CTHYB and found that the real part of the Matsubara self-energy is too high at nearly 20 as shown below, which deviate a lot from reference. I have adjusted some parameters of the CTHYB solver as you suggested above, but this condition didn't change. Could you please comment on this?
![image](https://github.com/TRIQS/solid_dmft/assets/51743630/cb0881bf-aafc-420f-b05c-366df2767426)
![image](https://github.com/TRIQS/solid_dmft/assets/51743630/fb321561-42ea-4066-aa9c-6d0164580815)
![image](https://github.com/TRIQS/solid_dmft/assets/51743630/31b88388-b49d-4074-ab7b-8b19b523076c)
![image](https://github.com/TRIQS/solid_dmft/assets/51743630/06b11f04-55e8-4abd-a595-2fb31205be29)

**Reply by the-hampel** [reply]
Hi @Lanlin123 , 
let's start with addressing the non convergence in FeSe and the output file you provided. Please be advised that it only makes sense to choose one tail treatment procedure: 
```
legendre_fit=True
measure_G_l=True
improved_estimator=True
```
You should only turn on of these flags on. Please use `lengendre_fit` first and then try other settings. Never turn more than one of these to `True`. You can see also a warning in the output: 
```
 !!!!WARNING!!!! 
 you enabled both improved estimators and legendre based filtering / sampling. Sigma will be overwritten by legendre result.  
 !!!!WARNING!!!!
 ```
Please check the output file for warnings. 

Now, it seems that you run into instabilities during the DMFT iterations. Your double counting potential fluctuates drastically. This could be because your number of legendre polynomials is too large. Did you check the convergence of Gl? Please rerun the calculation with `n_l=30` and check if results change. 

If this does not fix things: I noticed that `V_DC` in each iteration of the std out is changing rapidly. I suggest to perform first a calculation with `dc_dmft = False`. This fixes the DC in the first iteration. If this converges you can restart the calculation and allow the DC calculation to run normally in each DMFT iteration. 

Regarding memory issues I am not sure. How much memory per core does your worker nodes have? Typically less than one GB per core should be sufficient. Does this occur in the first DMFT iteration? Or only in the calculations where mu drifts so enormously?

**Reply by Lanlin123** [reply]
Hi @the-hampel,
Thanks for your reply. There is 3.7GB memory per code in my work machine. This crash just occured in the first DMFT iteration. By the way, I have no idea why the real part of the Matsubara self-energy of FeSe from CTHYB is so large. I think it's the reason that the pole of the AC real frequency self-energy near omega=0 point occurs as shown in the above cells.

**Reply by the-hampel** [reply]
I thought this was a result of the wildly oscillating DC potential in the ctseg calculation. Is the real part of Sigma also so large in the earlier cthyb calculations? 

3.7GB mem per core should be more than sufficient. Very weird. Does it happen in the first iteration before you see any `warmup` and `accumulation` output of the solver? (can you send me the output from that run).

Btw. you can also do calculations with fixes nominal occupations in solid_dmft if that is of interest. For this you have to add in the `[advanced]` section in the input `dc_fixed_occ= some value` and choose for the value whatever is the nominal occupation of the impurity in your model.

**Reply by Lanlin123** [reply]
Yeah, actually the Matsubara self-energy shown above is the result of cthyb calculation (without segment). I have changed the parameters as you suggeted above, i.e., using **perform_tail_fit = true, fit_max_moment = 4, fit_min_w = 3.0, fit_max_w = 8.0**. But it seems that this large real part is so robust.

The crash happens after just the first output of **warmup** and **accumulation** as shown below.
![image](https://github.com/TRIQS/solid_dmft/assets/51743630/cffd6d6d-721e-410b-9ab3-68e0ff8fbe9a)
 Thanks for your advice. I will try to using nominal occupations.

**Reply by the-hampel** [reply]
Hm, that indeed looks like a bug in the code or some issue upon collecting results. This is all still within the C++ part of the code, and I am not sure what causes the memory consumption here. Is this also happening with `measure_Giw=False` and `measure_Gl=False` ? Otherwise I have little ideas here why this happens.

**Reply by Lanlin123** [reply]
Hi @the-hampel,
The crash orginates from my wrong parameter settings. I set

> legendre_fit=True
> measure_G_l=True
> improved_estimator=True
at the same time. When I choose only **legendre_fit=True**, the crash disappears.

**Reply by the-hampel** [reply]
Ah good to know. Most likely the improved estimator is using quite a bit of memory. Glad it works now!

**Reply by Lanlin123** [reply]
Hi @the-hampel,
Unfortunately, there is still something wrong with my calculation of FeSe. I said before that the real part of the FeSe Matsubara self-energy is large. I found that this is because the occupation number of impurities is 6, resulting in a large value of DC potential, which seems reasonable. Both CTHYB and CTSEG give the similar result. I adjusted various input parameters to improve the convergence criteria, but the result didn't change. So I think the problem stems from analytic continuation, which results in a pole in both the real part and the imaginary part of real frequency self energy around w=0. I also tried various parameters of the postprocessing.maxent_sigma function and found that the ‘direct’ continuator_type behaves better, but the values of real frequency self energy around w=0 are still too large to give the reasonable k resolved spectral function A(k,w). I have put all the files in fese.rar which can be accessed from [https://drive.google.com/file/d/1QRDAFC_vQGz0k32ZmyNPQFF0J3hqq_SZ/view?usp=sharing](url). Could you please see them and give some advice? As FeSe is a typical material for DMFT, I wonder if it is convenient for your team to do this calculation and provide a tutorial. Thanks for your work.
 
Best regards,
Lanlin.

**Reply by the-hampel** [reply]
Hi, 
I had a look at your calculation. From a numerical point everything looks reasonable. The QMC seems to give good enough statistics and the DMFT convergence looks good. Let me stress that the relevant self-energy shift entering your spectral function is not simply the value of Sigma_iw. You have to plot something like this: 
```
with HDFArchive('fese.h5','r') as ar:
    print(ar['DMFT_results/last_iter'])
    S_iw = ar['DMFT_results/last_iter/Sigma_freq_0']
    mu_dft =  -0.00752
    mu_dmft = ar['DMFT_results/last_iter/chemical_potential_post']
    dc = ar['DMFT_results/last_iter/DC_pot'][0]['up'][0,0]

oplot(S_iw['up_0']+mu_shift-dc)
```
Explanation: The relevant part is in the lattice gf +mu- Eps_k - (Sigma - Sigma_DC), so the large shift you see in the real part of Sigma is mostly absorbed by the DC. So for example on the Matsubara axis this looks like this: 
![image](https://github.com/TRIQS/solid_dmft/assets/21337117/6c59d4c1-6582-4199-b5fa-8ae3177c7171)
and I also tried a quick Pade analytic continuation: 
![image](https://github.com/TRIQS/solid_dmft/assets/21337117/12cd0261-9b99-4bad-831d-ca66a4c64c91)
which looks comparable to what is happening on the Matsubara axis. Maybe you can check your MaxEnt data more closely (they were not included), but on the large scale in the plot they look reasonable. 

So from that point everything looks good. solid_dmft gave you the answer for these input parameters. I am not sure how to help beyond that? I agree that the self-energy difference between the orbitals red vs orange is definitely larger compared to what is shown the in the literature plot. But this is a result of the filling, input states, and interaction. If you want to exactly reproduce a self-energy from literature you would need to adjust the input projectors to be from larger energy window I guess, and tune interaction parameters. This is a bit out of scope for me, since all technical issues are resolved at this point as far as I can see. But given the strong mass enhancement of Sigma the resulting spectral function is reasonable to me.

The only thing I did not check if your initial lattice gf filling of 12 electrons is reasonable? And also I would try to not remove the dft fermi level in the w90 converter step. This makes post-processing only more complicate.

I added my post-processing parts here in the notebook you send me:
https://www.dropbox.com/scl/fi/85bmhf7g7juavu9fbxfvg/dojoctseg.ipynb?rlkey=l75runig6qbqgercr3ou8dgb2&dl=0

**Reply by Lanlin123** [reply]
Hi @the-hampel,
I see. Thank you for your patient and detailed answer. I will adjust the input projectors to be from larger energy window and tune interaction parameters. For the former one, I meet some problems in the test of SrVO3 tutorial. Firstly, I consider both V d orbitals and O p orbitals in my wannier caculation instead of only V d orbitals in the tutorial. Then, I choose bloch basis to generate the h5 archive. The problems include:

- When the O p orbitals is included, total density for imp 0 from DFT becomes 2.736551, deviating from 1.0, the total density for imp 0 when considering only V d orbitals. I doubt if there is something wrong in the process of determining impurity density for the bloch basis when considering uncorrelated orbitals.
- When the O p orbitals is included and there is 14 bands in the TB model, the dmft calculation will crash in the warmup process of  the first iteration using CTSEG solver. From the log file, It seems to be caused by insufficient memory. My mechine include 52 cores with totally 192 GB memory for regular node and 64 cores with 1.5 TB memory for bigmem node. But both didn't work. I think 5 correlated orbitals + 9 uncorrelated orbitals is not a Memory-intensive calculation for DMFT calcution. So this memory problem will severely limits the scope of my use of solid_dmft. Actually I've encountered this problem before in other calculations that considered uncorrelated electrons using CTHYB solver. Could you please give me some advice on this?

All the files can be accessed from [https://drive.google.com/file/d/1IanK4idv3PhyaRSUhp1kNMABgqdmPVDN/view?usp=drive_link](url). Thanks again.

**Reply by the-hampel** [reply]
Hi @Lanlin123 ,

In principle this looks all reasonable and you are absolutely right and there should be no memory problem here. I tried to run the calculation from your archive and the error that you are seeing is not a limit in memory. I definitely get in the error output (which I don't see in the output dir) the following error: 
```
Warning: The Delta(tau) block number 0 is not real in tau space                                                                                                                                     
Warning: The Delta(tau) block number 1 is not real in tau space                                                                                                                                     
                                                                                                                                                                                                    
Warming up ...                                                                                                                                                                                      
                                                                                                                                                                                                    
Intel MKL ERROR: Parameter 9 was incorrect on entry to DGER  .              
```
a quick look in the `svo_hr.dat` reveals the problem. You somehow have a Wannier Hamiltonian with large imaginary elements. This is not good and points to some problem during the wannierization. Before you can run DMFT you have to first try to Wannierize the system better, because the solver cannot handle imaginary hybridization functions. I noticed that you only used 100 wannierization steps in w90. Is there a reason? There is definitely a way to build Wannier functions without imaginary hoppings. You can try to use more kpoints or converge dft better first, but this is only wannier90 problem not something in solid_dmft. 

Best,
Alex

**Reply by Lanlin123** [reply]
Hi @the-hampel,
I can see some warning during the convert process like this:
![image](https://github.com/TRIQS/solid_dmft/assets/51743630/e9997579-1b9c-41f5-bd8a-8b1f987d038a)
But when I use more kpoints, i.e., from 8 8 8 to 15 15 15, and 3000 wannierization steps, this warning will disappear. And I have checked that the imaginary part of hoppings in hr_dat file is 0 now. However, the same crash occurs, like this:
![image](https://github.com/TRIQS/solid_dmft/assets/51743630/799edd43-91ed-4ad8-aca9-3a0511c974f4)
It seems that the more converged wannier result does not make the Delta(tau) become real. And the impurity density get by bloch basis is still 2.7 but not 1, which seems something goes wrong.
Then I made a test using wannier basis including O p orbitals. In this case, if I set the electron density as 19 in _svo.inp_, the impurity density recognized by the program is 2.7 too. And the crash occurs the same as the bloch basis case. But when I change the electron density from 19 to 1 in _svo.inp_ , the impurity density recognized by the program is about 0.1 and the calculation can go through successfully with the same warning "_Warning: The Delta(tau) block number 0 is not real in tau space_". So this suggest that the cause of the program crash is a problem with charge density recognition when considering the uncorrelated O p orbitals. Could you please help on this?
Apologies for keep disturbing you and your team. Your help is always appreciated!

**Reply by the-hampel** [reply]
Okay so let's discuss a bit the part about the occupations. The change you see in occupations is totally reasonable. You are right that in principle there are three t2g like bands at the Fermi level that are filled 1/6, i.e. one electron in SrVO3. However, in pd-models where you include oxygens the lattice Greens function is now including all the oxygen states (3 x 3 orbitals times 2 for spin) which can holds 18 electrons and are completely filled. This is why you end up with 19 electrons in the lattice Gf. No extracting the local Green's function of the d electrons is now not giving you 1 electron anymore, because the orbitals are not purely made out of the 3 t2g like states but rather atomic like d orbitals that mix with the oxygen. Hence, your orbitals also have some occupation stemming from the O-p orbitals. I think 2.7 is totally reasonable. And the fact that you get the same number in bloch and wannier basis is very good. 

Can you take the large kpoint calculation, check that there are no imaginary numbers in `svo_hr.dat`, and then run solid_dmft using the cthyb solver one DMFT iteration. If that works we go back to ctseg.

**Reply by Lanlin123** [reply]
Hi @the-hampel,
I have used cthyb solver, and it do work. But the calculation speed is a bit slow. The _Accumulating_ process hasn't finished yet. Files can be accessed from [https://github.com/Lanlin123/test-for-SrVO3](url).  As the warmup process was successfully completed, which is different from the ctseg solver case, I think the cthyb sover do work. The problem is with ctseg solver.

**Reply by the-hampel** [reply]
Hi, the output now looks really good with respect to imaginary elements etc. Nice. Now let's try to understand why ctseg fails. I cloned your repo that you shared and played around with it. The problem is that you try to run the impurity solver with all off-diagonal terms in the hybridization (`enforce_off_diag=True`), but thanks to your nice wannierization your problem is really diagonal. This you can see in the output of the local Hamiltonian (H_loc0 and rho) in the top. In cthyb there is default cut-off for off-diagonal terms, something like 10^-12 or so, which probably removes the off-diagonal components by default in the solver. However, the ctseg solver does not have this. To help the solver do the following in the input:
```
enforce_off_diag = False
block_threshold= 1e-5   
```
This will give you now a more diagonal block_structure:
```
 | block struct. solver: {'up_0': 1, 'up_1': 1, 'up_2': 1, 'up_3': 1, 'up_4': 1, 'down_0': 1, 'down_1': 1, 'down_2': 1, 'down_3': 1, 'down_4': 1}
  | deg. orbitals : [['up_0', 'up_1', 'up_2', 'down_0', 'down_1', 'down_2'], ['up_3', 'up_4', 'down_3', 'down_4']]
  ```
  Correctly identifying the t2g and eg as degenerate. With this change in the input file ctseg runs for me (before I could reproduce your problem). Additionally consider: 
  ```
  measure_G_iw=False
  ```
  That should improve speed further and does not help you much. The setting for the block_structure has always done with care. You have to check afterwards if the correct block structure and degeneracies are detected. This is why by default it is not done.

Let me know if this fixes your problem.

**Reply by Lanlin123** [reply]
Hi @the-hampel ,
Sorry for the late reply as there are too many things going on recently. I add the tag **enforce_off_diag=True** in the ctseg solver caculation and it works. The correct block structure and degeneracies are detected. I’m very grateful to you and your team for helping me over a long period of time.

Best regards,
Lanlin.

**Reply by the-hampel** [reply]
You are very welcome - happy to hear that it works now.

**Comment by phibeck** [comment]
> Hi @Lanlin123, this seems to be a bug in the Wannier90Converter. I've opened an issue in https://github.com/TRIQS/solid_dmft/issues/76. We'll try to address this as soon as possible, thanks for providing input files.

Hi @Lanlin123 the bug you reported should be fixed in the unstable branch of DFTTools. https://github.com/TRIQS/dft_tools/pull/252. You can recompile at it should be gone, or alternatively you can set a disentanglement window (you can make it as large as you want) and that should fix it, too. Let us know if you encounter any more issues.

**Reply by Lanlin123** [reply]
Hi @phibeck,
Yeah, I see. Thanks for your works.
